<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [75]:
import pandas as pd
import psycopg2

In [76]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [77]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [78]:
# текст запроса
query_3_1 = f'''select count(name)
           from public.vacancies
'''

In [79]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [80]:
# текст запроса
query_3_2 = f'''select count(name)
           from public.EMPLOYERS
'''

In [81]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [82]:
# текст запроса
query_3_3 = f'''select count(name)
           from public.AREAS
'''

In [83]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [84]:
# текст запроса
query_3_4 = f'''select count(name)
           from public.INDUSTRIES
'''

In [85]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

Выводы по предварительному анализу данных

База данных содержит информацию о порядка 50 тысячах вакансиях и 23,5 тысячах работодателей в 1 362 регионах России, в базе представлены 294 сферы деятельности (количество уникальных вакансий, работадателей, сфер деятельности и регионов на данном этапе не рассчитывалось, данные представлены в абсолютных величинах).

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [86]:
# текст запроса
query_4_1 = f'''select count(v.name) as cnt, a.name
           from public.areas as a
           join public.vacancies as v on v.area_id=a.id
           group by 2
           order by 1 desc
'''

In [87]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt             name
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [88]:
# текст запроса
query_4_2 = f'''select count(id)
           from public.VACANCIES
           where salary_from is not null or salary_to is not null
'''

In [89]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [90]:
# текст запроса
query_4_3 = f'''select round(avg(salary_from),0), round(avg(salary_to),0)
           from public.VACANCIES
'''

In [91]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round     round
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [92]:
# текст запроса
query_4_4 = f'''select schedule, count(schedule), employment, count(employment)
           from public.VACANCIES
           group by schedule, employment
           order by 2 desc, 4 desc
'''

In [93]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule  count           employment  count
0        Полный день  35367     Полная занятость  35367
1   Удаленная работа   7802     Полная занятость   7802
2      Гибкий график   1593     Полная занятость   1593
3   Удаленная работа   1312  Частичная занятость   1312
4     Сменный график    940     Полная занятость    940
5        Полный день    569           Стажировка    569
6     Вахтовый метод    367     Полная занятость    367
7        Полный день    347  Частичная занятость    347
8      Гибкий график    312  Частичная занятость    312
9        Полный день    141     Проектная работа    141
10  Удаленная работа    133     Проектная работа    133
11     Гибкий график    116           Стажировка    116
12    Сменный график    101  Частичная занятость    101
13  Удаленная работа     64           Стажировка     64
14     Гибкий график     18     Проектная работа     18
15    Сменный график     12           Стажировка     12
16    Вахтовый метод      2     Проектная работа      2
17    Сменный график      1     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [94]:
# текст запроса
query_4_5= f'''select experience, count(id) cnt
           from public.VACANCIES
           group by experience
           order by 2
'''

In [95]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

Выводы по детальному анализу вакансий

Топ-5 регионов по количеству представленных вакансий: Москва, Санкт-Петербург, Минск, Новосибирск и Алматы. Наименьшее количество вакансий прдставленно в регионе Анна - 1. Менее, чем у половины вакансий (24 тысячи из 49 тысяч) заполнено хотя бы одно из двух полей о заработной плате (верхняя/нижняя граница). В среднем зарплатная вилка находится в промежутке от 71 тысячи до 110,5 тысяч рублей. Наиболее распространенным является сочетания типа рабочего графика и типа трудоустройства: полный день, полная занятость. Наиболее часто встречающееся требование к опыту работы - от 1 года до 3 лет (26 тысяч вакансий), наиболее редкое - более 6 лет (1 тысяча вакансий).

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [96]:
# текст запроса
query_5_1= f'''select e.name, count(v.id) cnt
           from public.EMPLOYERS e
           join public.vacancies v on v.employer_id=e.id
           group by 1
           order by 2 desc
           limit 5
'''

In [97]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [98]:
# текст запроса
query_5_2= f'''select a.name area, count(e.id) employers, count(v.id) vacancies
           from public.areas a
           left join public.employers e on a.id=e.area
           left join public.vacancies v on a.id=v.area_id
           group by 1
           order by 3, 2 desc
'''

In [99]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area  employers  vacancies
0                 Россия        410          0
1              Казахстан        207          0
2     Московская область         75          0
3     Краснодарский край         19          0
4     Ростовская область         18          0
...                  ...        ...        ...
1357         Новосибирск    1149438    1149438
1358              Алматы    1364132    1364132
1359               Минск    2354880    2354880
1360     Санкт-Петербург    6320667    6320667
1361              Москва   31272712   31272712

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [100]:
# текст запроса
query_5_3= f'''select v.employer_id, count(distinct v.area_id) cnt
           from public.vacancies v
           group by 1
           order by 2 desc
'''

In [101]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer_id  cnt
0             1740  181
1             2748  152
2          5724811  116
3          5130287   88
4          3682876   71
...            ...  ...
14901       810278    1
14902       810313    1
14903       810551    1
14904       810688    1
14905       811902    1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [102]:
# текст запроса
query_5_4= f'''select count(e.id) cnt
           from public.employers e
           left join public.employers_industries ei on ei.employer_id=e.id
           group by employer_id
           having employer_id is null
'''

In [103]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [104]:
# текст запроса
query_5_5= f'''select e.name
           from public.employers e
           join public.employers_industries ei on ei.employer_id=e.id
           group by e.name
           having count(industry_id)=4
           order by 1
           limit 3
'''


In [105]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name
0  101 Интернет
1      21vek.by
2          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [106]:
# текст запроса
query_5_6= f'''select count(i.name) cnt
           from public.EMPLOYERS_INDUSTRIES ei
           join public.INDUSTRIES i on ei.industry_id=i.id
           join public.employers e on ei.employer_id=e.id
           group by i.name
           having i.name='Разработка программного обеспечения' 
'''

In [107]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [108]:
# код для получения списка городов-милионников
import requests
tables = pd.read_html(url)
city_df = tables[1]
city_list = list(city_df['Город'])
city_list[13] = 'Воронеж'
print(city_list)

NameError: name 'url' is not defined

In [ ]:
# текст запроса
query_5_7= f'''select a.name, count(v.id) cnt
           from public.vacancies v
           join public.employers e on e.id=v.employer_id and e.name='Яндекс'
           join public.areas a on a.id=v.area_id and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
           group by a.name
           union all
           select 'Total', count(*)
           from public.vacancies v
           join public.employers e on e.id=v.employer_id and e.name='Яндекс'
           join public.areas a on a.id=v.area_id and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
           order by cnt
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

Выводы по анализу работодателей

Наибольшее количество выставленных вакансий принадлежит Яндексу. Далее идут Рсотелеком, Тинькофф, СБЕР и Газпром нефть. Наибольшее количество вакансий и работодателей представлено в Москве, тогда как в таком регионе как Россия не представлено ни одной вакансии (как можно заметить по результату запроса, работодатели указывают определенный населенный пункт, а не непосредственно страну/область). У почти 8,5 тысяч работодателей не указана сфера деятельности, в которой они работают. Чуть более, чем 3,5 тысячи работодателей работабт в сфере Разработки программного обеспечения. Вакансии компании Яндекс представлены в 16 городах-милиионниках, общее количество вакансий в этих регионах составляет 485 штук, лидирующими решионами являются Москва (54 вакансии) и Санкт-Петербург (42 вакансии), наименьшее количество вакансий представлено в Омске (21 вакансия). 

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса
query_6_1= f'''select count(v.name) cnt
           from public.VACANCIES v
           where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса
query_6_2= f'''select count(v.name) cnt
           from public.VACANCIES v
           where (lower(v.name) like '%data%scientist%' 
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%'
                or v.name like '%ML%'
                and v.name not like '%HTML%')
                and (lower(v.name) like '%junior%'
                or lower(v.experience) like '%нет%опыта%'
                or (v.employment) like '%cтажировка%')
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3= f'''select count(v.name) cnt
           from public.VACANCIES v
           where (lower(v.name) like '%data%scientist%' 
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%'
                or v.name like '%ML%'
                and v.name not like '%HTML%')
                and (v.key_skills like '%SQL%'
                or lower(v.key_skills) like '%postgres%')
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4= f'''select count(v.name) cnt
           from public.VACANCIES v
           where (lower(v.name) like '%data%scientist%' 
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%'
                or v.name like '%ML%'
                and v.name not like '%HTML%')
                and lower(v.key_skills) like '%python%'
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса
query_6_5= f'''select round(avg(length(v.key_skills)-length(replace(v.key_skills, CHR(9), ''))+1),2) avg
           from public.VACANCIES v
           where lower(v.name) like '%data%scientist%' 
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%'
                or v.name like '%ML%'
                and v.name not like '%HTML%'
                and v.key_skills is not NULL
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avg
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [117]:
query_6_6= f'''select distinct v.experience, round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) avg_salary
           from public.VACANCIES v
           where lower(v.name) like '%data%scientist%' 
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%'
                or v.name like '%ML%'
                and v.name not like '%HTML%'
            group by v.experience
'''

In [118]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

c:\Users\kslat\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  avg_salary
0         Более 6 лет         NaN
1           Нет опыта     74643.0
2  От 1 года до 3 лет    139675.0
3       От 3 до 6 лет    243115.0

Выводы по предметному анализу

Всего в базе представлено 1 771 вакансия, имеющая отношение к данным. Из них подходящих вакансий для начинающих дата-саентистов 51. В целом для дата-саентистов требование ключевого навыка знание SQL или postgres - 201 (порядка 10% от количества вакансий, относящихся к данным). Python более популярен среди требований - 351 вакансия (почти 20% вакансий). В среднем работодатели указывают порядка 6 ключевых требований к соискателям. Средняя заработная плата, на которую может рассчитывать начинающий специалист (т.е. от с опытом работы от 1 года до 3 лет) - около 140 тысяч, специалист с более большим опытом работы (от 3 до 6 лет) - порялка 243 тысяч. Информация о вилке заработной платы по опытным специалистам не указана.

***

# Общий вывод по проекту

In [2]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

База данных содержит информацию о порядка 50 тысячах вакансиях, однако только 1 771 вакансия имеет отношение к данным. Из них подходящих вакансий для начинающих дата-саентистов 51. В целом для дата-саентистов требование ключевого навыка знание SQL или postgres - 201 (порядка 10% от количества вакансий, относящихся к данным). Python более популярен среди требований - 351 вакансия (почти 20% вакансий). В среднем работодатели указывают порядка 6 ключевых требований к соискателям. Средняя заработная плата, на которую может рассчитывать начинающий специалист (т.е. от с опытом работы от 1 года до 3 лет) - около 140 тысяч, специалист с более большим опытом работы (от 3 до 6 лет) - порялка 243 тысяч. Информация о вилке заработной платы по опытным специалистам не указана.

Наибольшее количество выставленных вакансий принадлежит Яндексу, также наибольшее количество вакансий и работодателей представлено в Москве. Чуть более, чем 3,5 тысячи работодателей работают в сфере Разработки программного обеспечения. Например, вакансии компании Яндекс представлены в 16 городах-милиионниках. Так как деятельность компании направлена на работу с большим количеством информации, рекомендациями (например, Яндекс.Музыка), можно пройти стажировку, а офисы представлены почти во всех крупнейших городах страны, то молодым специалистам стоит обратить внимание в том числе и на данного работодателя.